overlord/state/state_test.go

```diff
@@ -1081,3 +1081,36 @@ func (ss *stateSuite) TestNoStateErrorString(c *C) {
 	err.Key = "foo"
 	c.Assert(err.Error(), Equals, `no state entry for key "foo"`)
 }
+
+func add_cache(st *state.State, c *C, what string) {
+	var m map[string]int
+	if x := st.Cached("M"); x != nil {
+		m = x.(map[string]int)
+		c.Assert(m, NotNil)
+	} else {
+		m = make(map[string]int)
+	}
+	m[what] = 1
+	st.Cache("M", m)
+}
+
+func remove_cache(st *state.State, what string) {
+	m, _ := st.Cached("M").(map[string]int)
+	delete(m, what)
+	if len(m) == 0 {
+		m = nil
+	}
+	st.Cache("M", m)
+}
+
+func (ss *stateSuite) TestCacheDeleteMap(c *C) {
```

Comment:
why the test here? we need a test in snapstate to show the issue there

Reply:
Because the bug is, in fact, in state: the `Cache()` method is the one that must discern if the passed interface points to nil.

Reply:
Unless you prefer me to test the functions at handlers.go...

Reply:
Ok, removed all that.